.github/CONTRIBUTING.md

```diff
@@ -1,42 +1,251 @@
+* Use return early pattern. [See more](https://blog.timoxley.com/post/47041269194/avoid-else-return-early)
+* Prefer `Promise` over `callbacks`
+* Prefer `await` over `then/catch`
+* Don't create queries outside models, the query description should be inside the model class.
+* Don't hardcode fields inside models. Same method can be used for different purposes, using different fields.
+* Prefer create REST endpoints over Meteor methods
+* Prefer call REST endpoints over Meteor methods when both are available
+* v1 REST endpoints should follow the following pattern: `/api/v1/dashed-namespace.camelCaseAction`
+* Prefer TypeScript over JavaScript. Check [ECMAScript vs TypeScript](#ecmascript-vs-typescript)
 
-We acknowledge all the code does not meet these standards but we are working to change this over time.
+#### Blaze
+* Import the HTML file from it's sibling JS/TS file
 
 ### Syntax check
 
 Before submitting a PR you should get no errors on `eslint`.
 
-To check your files, first install `eslint`:
+To check your files run:
+
+```shell
+meteor npm run lint
+```
+
+## Tests
+
+There are 2 types of tests we run on Rocket.Chat, **Unit** tests and **End to End** tests. Different from the Unit tests the End to End depends on a Rocket.Chat instance running to execute the API and UI checks.
+
+### End to End Tests
 
+First you need to run a Rocket.Chat server on **Test Mode** and on a **Empty Database**:
+```shell
+# Running with a local mongodb database
+MONGO_URL=mongodb://localhost/empty MONGO_OPLOG_URL=mongodb://localhost/local TEST_MODE=true meteor
 ```
-npm install -g eslint
+```shell
+# Running with a local mongodb database but cleaning it before
+mongo --eval "db.dropDatabase()" empty && MONGO_URL=mongodb://localhost/empty MONGO_OPLOG_URL=mongodb://localhost/local TEST_MODE=true meteor
 ```
 
-Then run:
+Now you can run the tests:
+```shell
+meteor npm test
+```
+
+### Unit Tests
 
+Unit tests are simple to run, the don't depends on a Rocket.Chat instance running, just execute:
+```shell
+meteor npm run testunit
 ```
-eslint .
+
+It's possible to run on watch mode as well:
+```shell
+meteor npm run testunit-watch
 ```
 
-# Contributor License Agreement
+<!-- ### Storybook -->
+
+## Before Push your code
+
+It's important to run the lint and tests before push your code or submit a Pull Request, otherwise your contribution may fail quickly on the CI and the reviewers will need to request you to fix those errors delaying the review of your contribution.
+
+Rocket.Chat uses [husky](https://www.npmjs.com/package/husky) to run the **lint** and **unit tests** before proceed to the code push process, so you may notice a delay when pushing your code to your repository.
+
+## Choosing a good PR title
+
+We use PR titles to on our change log, so when writing the PR title bare in mind it should make sense to a person reading the a release's change log.
```

Comment:
Suggestion:

It is very important to note that we use PR titles when creating our change log.  Keep this in mind when you title your PR.   Make sure the title makes sense to a person reading a releases's change log!